compiler/rustc_mir_transform/src/remove_uninit_drops.rs

```diff
@@ -102,7 +102,7 @@ fn is_needs_drop_and_init<'tcx>(
     let field_needs_drop_and_init = |(f, f_ty, mpi)| {
         let child = move_path_children_matching(move_data, mpi, |x| x.is_field_to(f));
         let Some(mpi) = child else {
-            return f_ty.needs_drop(tcx, param_env);
+            return Ty::needs_drop(f_ty, tcx, param_env);
```

Comment:
Without this this code does not compile because the type of `f_ty` is not resolved to `Ty`.

Here is a reason why this compiled with lowering in AST but not anymore with lowering in MIR. With AST lowering, this let...else was transformed to:
```rust
if let Some(mpi) = child {
    is_needs_drop_and_init(tcx, param_env, maybe_inits, move_data, f_ty, mpi)
} else {
    return f_ty.needs_drop(tcx, param_env);
}
```
Since the `is_needs_drop_and_init` function call was visted first, typeck resolved the type of `f_ty` to `Ty` before visiting the return statement.

We no longer visit the expressions in this order with lowering MIR. This time, the return statement is visited first. We need to resolve the method call but `f_ty` has no type obligations at this point yet.

Reply:
After the adjustment in order, even after moving the check on the `else` block into `check_decl` at the tail position, this is unfortunately still necessary as `type annotation is needed` is still emitted with the unchanged call for the reason possibly presented in the previous comment.

Reply:
oh 🤦 now I see it. Thanks!

I guess I'd prefer to annotate the `f_ty` argument instead of changing this call site.